language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -127,6 +139,113 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
         existing_borrows.is_empty()
     }
 
+    fn verify_instantiation(
+        env: (&[StructHandle], &[Kind], usize),
+        tys: &[SignatureToken],
+        type_formals: &[Kind],
+    ) -> Result<Vec<Kind>, VMStaticViolation> {
+        let (_, _, offset) = env;
+
+        // Check if the number of formals and the number of actuals match.
+        if type_formals.len() != tys.len() {
+            return Err(VMStaticViolation::NumberOfTypeActualsMismatch(
+                offset,
+                type_formals.len(),
+                tys.len(),
+            ));
+        }
+
+        // Recursively verify the type actuals and gather their kinds.
+        let kinds = tys
+            .iter()
+            .map(|ty| Self::verify_type(env, ty))
+            .collect::<Result<Vec<_>, VMStaticViolation>>()?;
+
+        // Check if each kind is a sub-kind of that specified in the constraint.
+        if kinds
+            .iter()
+            .zip(type_formals.iter())
+            .any(|(k1, k2)| !k1.is_sub_kind_of(*k2))
+        {
+            return Err(VMStaticViolation::KindMismatch(offset));
+        }
+
+        Ok(kinds)
+    }
+
+    fn verify_type(
```

Comment:
This function does two things:
1) Verifying that the types & kinds are good.
2) Determining the kind of a given type. This can be non-trivial as we have type actuals now.

Now here's one awkwardness: in some places we just need to get the kind of a type and don't require constraints to be checked for instantiation. However we still need to collect the kinds of the type actuals and do all other work as we do in `verify_type`. I'm wondering if it makes sense to write another `fn get_kind` or can we just use `verify_type`. If we do, we'll end up with a bit of duplicated code. Macros can help, but I'm not sure if that's worth it.

Reply:
Also I decided not to fix `fn kind` in `vm/src/views.rs` (for now).
1) As shown by my current implementation, verifying the correctness of kinds & getting the kinds of a type are pretty much the same problem or at least closely related. If you solve one you pretty much have solved another one. (The "awkwardness" I mentioned above.)
2) Currently `SignatureTokenView` only knows the module it is in (it doesn't even support scripts!), but to determine the kind of a type parameter we need to know the type formals, or in other words, the function/struct that contains the type parameter. In other words, we need to have access to what I would call a "local scope". There are two approaches I can think of to fix it:
  - Rewrite views so that `SignatureTokenView` has access to a `LocalScope` and the `LocalScope` has access to a `GlobalScope` (module/script). This can be done but it's not my top priority.
  - Add a field to `SignatureTokenView` so it has access to the `LocalScope`. The problem with this is that if you modify `SignatureTokenView`, you also need to change other structs like `TypeSignatureView` and `FieldDefinitionView`, which results in many call sites to fix. I think this is a distraction for now and I would focus on fixing the verifier.